lib/posix-netlink/netlink.c

```
@@ -0,0 +1,309 @@
+	nl_ctx->bound = 1;
+	return 0;
+}
+
+static int nl_sockname(posix_sock *s,
+		       struct sockaddr *restrict addr,
+		       socklen_t *restrict addr_len)
+{
+	struct sockaddr_nl *nl_addr;
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+
+	if (*addr_len < sizeof(*nl_addr)) {
+		/* addr_len is too small, only inform about needed size */
+		goto out;
+	}
+
+	nl_addr = (struct sockaddr_nl *)addr;
+	nl_addr->nl_family = AF_NETLINK;
+	nl_addr->nl_pid = nl_ctx_pid(nl_ctx);
+	nl_addr->nl_pad = 0;
+	nl_addr->nl_groups = nl_ctx->nl_groups;
+
+out:
+	*addr_len = sizeof(*nl_addr);
+	return 0;
+}
+
+static inline int nl_handle(struct nl_ctx *nl_ctx, const void *buf, size_t len)
+{
+	const struct nlmsghdr *nlh;
+	size_t tmp_len;
+	int ret = 0;
+
+	UK_ASSERT(nl_ctx);
+	UK_ASSERT(nl_ctx->drv);
+	UK_ASSERT(nl_ctx->drv->ops);
+	UK_ASSERT(nl_ctx->drv->ops->handle);
+
+	uk_pr_debug("Handle incoming netlink msg\n");
+	if (!buf || len < sizeof(struct nlmsghdr))
+		return -EINVAL;
+
+	/* sanity check multi-messages */
+	tmp_len = len;
+	for (nlh = (const struct nlmsghdr *)buf;
+	     NLMSG_OK(nlh, tmp_len);
+	     nlh = NLMSG_NEXT(nlh, tmp_len)) {
+		if (tmp_len < nlh->nlmsg_len) {
+			uk_pr_debug("Sanity check failed\n");
+			return -EINVAL;
+		}
+	}
+
+	/* forward message by message */
+	for (nlh = (const struct nlmsghdr *)buf;
+	     NLMSG_OK(nlh, len);
+	     nlh = NLMSG_NEXT(nlh, len)) {
+		uk_pr_debug("Call handler for msg %p (driver %s)\n",
+			    nlh, nl_ctx->drv->libname);
+		ret = nl_ctx->drv->ops->handle(nl_ctx, nlh);
+	}
+
+	return ret;
+}
+
+static ssize_t nl_sendto(posix_sock *s,
+			 const void *buf, size_t len, int flags __unused,
+			 const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	struct nl_ctx *nl_ctx = sock2nlctx(s);
+	struct sockaddr_nl *nl_addr;
+	int ret;
+
+	if (addrlen < sizeof(*nl_addr))
+		return -EINVAL;
+
+	nl_addr = (struct sockaddr_nl *)dest_addr;
+	if (unlikely(nl_addr->nl_family != AF_NETLINK || nl_addr->nl_pid != 0))
```

`sendto` does not return `EHOSTUNREACH` according to manual... but it does according to POSIX :exploding_head: 
But I think the error here should be:
```C
if (unlikely(nl_addr->nl_family != AF_NETLINK))
	return -EAFNOSUPPORT;
if (unlikely(nl_addr->nl_pid != 0))
	return -EINVAL;  // or ECONNREFUSED
```

---

I think this is also one case where unikraft is special; in my (possibly naive) understanding of netlink as a _best effort_ protocol, sending a message to a destination that's not listening should silently drop the packet.
Thus in theory we should _not_ error out here, rather drop the request immediately (since in the current impl there cannot be anyone but the kernel listening). However, I'd rather have it throw an error to notify us of this missing feature, than have it stay silent, however correct the latter may be.
Again as for `bind`, error is up to us: `EINVAL` is already too overloaded, `ECONNREFUSED` is not mentioned in the `sendto(2)` manpage (nor does it technically apply here, since netlink sockets are not connection-oriented), so `EHOSTUNREACH` seems reasonable to me (esp since POSIX _does_ document it for `sendto`).